File: problems-adapt/0142_cycle-entry-in-a-list/solution_hash.go
type cycleListNode struct {
	val  int
	next *cycleListNode
}

func listCycleEntry(values []int, tailLink int) int {
	if len(values) == 0 {
		return -1
	}
	// Materialize the wire form: one node per value, then close the cycle.
	nodes := make([]*cycleListNode, len(values))
	for i, v := range values {
		nodes[i] = &cycleListNode{val: v}
	}
	for i := 0; i+1 < len(nodes); i++ {
		nodes[i].next = nodes[i+1]
	}
	if tailLink != -1 {
		nodes[len(nodes)-1].next = nodes[tailLink]
	}
	// Walk from the head remembering every node by address. The first node
	// to come around a second time is the cycle's entry; running off the
	// end instead means no cycle.
	seen := make(map[*cycleListNode]bool)
	node := nodes[0]
	for node != nil && !seen[node] {
		seen[node] = true
		node = node.next
	}
	if node == nil {
		return -1
	}
	// The judge wants an index: count steps from the head to the entry.
	index := 0
	entry := nodes[0]
	for entry != node {
		entry = entry.next
		index++
	}
	return index
}

File: problems-adapt/0000-0100/0054_cycle-entry-in-a-list/solution_hash.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def listCycleEntry(self, values: list[int], tailLink: int) -> int:
        if not values:
            return -1
        # Materialize the wire form: one node per value, then close the cycle.
        nodes = [ListNode(value) for value in values]
        for i in range(len(nodes) - 1):
            nodes[i].next = nodes[i + 1]
        if tailLink != -1:
            nodes[-1].next = nodes[tailLink]
        # Walk from the head remembering every node by identity. The first
        # node to come around a second time is the cycle's entry; running
        # off the end instead means no cycle.
        seen = set()
        node = nodes[0]
        while node is not None and node not in seen:
            seen.add(node)
            node = node.next
        if node is None:
            return -1
        # The judge wants an index: count steps from the head to the entry.
        index = 0
        entry = nodes[0]
        while entry is not node:
            entry = entry.next
            index += 1
        return index
